### ISSUE TITLE
Potential security concern when running in local mode (for development purposes)

### ISSUE THREAD

**Issue by parthea** [GoogleCloudPlatform/datalab#786]
Developers should be aware that by default Docker will expose the container port on any ip on the host machine. This is a greater concern if a developer has not setup a firewall (eg. iptables) and does not fully trust the network they are on (eg. public wifi ). Since authentication is disabled in local mode, anyone on the local network can connect using the datalab web interfance and run queries against the user's google account (since the .gcloud folder which contains authentication information is mounted from the host machine in the local option). 

I propose the following:

1. Add a security disclaimer in the wiki and in `$REPO_DIR/containers/datalab/run.sh`
and reference the solutions to restrict access mentioned in [Docker Advanced Networking](https://docs.docker.com/v1.5/articles/networking/)

2.  Modify the file `$REPO_DIR/containers/datalab/run.sh` to run docker with  `-p 127.0.0.1:host_port:container_port` instead of `host_port:container_port` 

From  [Docker Advanced Networking](https://docs.docker.com/v1.5/articles/networking/),

> You can see that Docker has exposed these container ports on 0.0.0.0, the wildcard IP address that will match any possible incoming port on the host machine. If you want to be more restrictive and only allow container services to be contacted through a specific external interface on the host machine, you have two choices. When you invoke docker run you can use either -p IP:host_port:container_port or -p IP::port to specify the external interface for one particular binding.

> Or if you always want Docker port forwards to bind to one specific IP address, you can edit your system-wide Docker server settings (on Ubuntu, by editing DOCKER_OPTS in /etc/default/docker) and add the option --ip=IP_ADDRESS. Remember to restart your Docker server after editing this setting.

**Comment by gramster** [comment]
Local mode is changing to not use mounted gcloud info. This will still be an issue though. Eventually a running kernel will have associated credentials and if someone opens an existing running session on a different machine they could have access. Locking down the ports may be the only good solution.

**Comment by gramster** [comment]
I've been trying -p 127.0.0.1 but haven't managed to make it work yet. Not sure if that is due to other corp restrictions we have; need to test it at home.

**Comment by parthea** [comment]
I just tested it again and `-p 127.0.0.1:8081:8080` worked but I was not up to date with master. Here is a diff of run.sh.

```
tony@TonyLinuxDesktop:~/datalab-parthea/containers/datalab$ git diff run.sh
diff --git a/containers/datalab/run.sh b/containers/datalab/run.sh
index 50651db..cadb560 100755
--- a/containers/datalab/run.sh
+++ b/containers/datalab/run.sh
@@ -42,7 +42,7 @@ PROJECT_NUM=`gcloud -q alpha projects describe $PROJECT_ID | g
 #  -v $REPO_DIR/sources/web:/sources \
 
 docker run -i --entrypoint=$ENTRYPOINT \
-  -p 8081:8080 \
+  -p 127.0.0.1:8081:8080 \
   -v $HOME/datalab/log:/var/log/app_engine \
   -v $HOME/.config/gcloud:/root/.config/gcloud \
   -v $REPO_DIR/content:/content \
```

After the update to the latest master, I'm having an issue compiling web sources. I'll try to setup the dev environment again to see if that helps.
```
tony@TonyLinuxDesktop:~/datalab-parthea/sources/web$ ./build.sh 
sending incremental file list

sent 80 bytes  received 12 bytes  184.00 bytes/sec
total size is 1,136  speedup is 12.35
sending incremental file list
extensions/

sent 621 bytes  received 19 bytes  1,280.00 bytes/sec
total size is 112,704  speedup is 176.10
sending incremental file list

sent 94 bytes  received 12 bytes  212.00 bytes/sec
total size is 22,550  speedup is 212.74
sending incremental file list

sent 63 bytes  received 12 bytes  150.00 bytes/sec
total size is 454  speedup is 6.05
datalab/static/extensions/charting.ts(428,16): error TS2339: Property 'innerHTML' does not exist on type 'Element'.
```

**Comment by parthea** [comment]
> I've been trying -p 127.0.0.1 but haven't managed to make it work yet. Not sure if that is due to other corp restrictions we have; need to test it at home.

@gramster I'm just following up to see if you've had success using  `-p 127.0.0.1:8081:8080` in your corporate network. Did you have to make any changes to get it to work?

cc @qimingj

**Comment by parthea** [comment]
Closing. This was fixed in #850
